webui/react/src/modal/useProjectEditModal.tsx

```diff
@@ -0,0 +1,99 @@
+import { FormInstance } from 'antd';
+import React, { useEffect, useMemo } from 'react';
+
+import Form from 'components/kit/Form';
+import Input from 'components/kit/Input';
+import { patchProject } from 'services/api';
+import { ErrorLevel, ErrorType } from 'shared/utils/error';
+import { Project } from 'types';
+import handleError from 'utils/error';
+import { observable, useObservable } from 'utils/observable';
+
+import { BaseModalState, ModalBody, useModal, UseModalParams, UseModalReturn } from './useModal';
+import css from './useProjectEditModal.module.scss';
+
+const FORM_ID = 'edit-project-form';
+
+interface FormInputs {
+  description?: string;
+  name: string;
+}
+
+interface ProjectModalState {
+  form: FormInstance<FormInputs>;
+  project?: Project;
+}
+
+type ModalState = ProjectModalState & BaseModalState;
+
+const ModalBody: ModalBody<ModalState> = ({ modalState }) => {
+  const { form, project } = useObservable(modalState);
+  if (!project) return null;
+  return (
+    <Form autoComplete="off" className={css.base} form={form} id={FORM_ID} layout="vertical">
+      <Form.Item
+        initialValue={project.name}
+        label="Project Name"
+        name="projectName"
+        rules={[{ message: 'Project name is required', required: true }]}>
+        <Input maxLength={80} />
+      </Form.Item>
+      <Form.Item initialValue={project.description} label="Description" name="description">
+        <Input />
+      </Form.Item>
+    </Form>
+  );
+};
+
+const handleSubmit = async (state: ModalState) => {
+  const { form, project } = state;
+  if (!project) return;
+  const values = await form.validateFields();
+  const { name, description } = values;
+
+  try {
+    await patchProject({ description, id: project.id, name });
+  } catch (e) {
+    handleError(e, {
+      level: ErrorLevel.Error,
+      publicMessage: 'Please try again later.',
+      publicSubject: 'Unable to edit project.',
+      silent: false,
+      type: ErrorType.Server,
+    });
+  }
+};
+
+const useProjectEditModal = (): UseModalReturn<ModalState> => {
+  const [form] = Form.useForm<FormInputs>();
+  const projectName = Form.useWatch('projectName', form);
+
+  const modalState = useMemo(
+    () => observable({ form, open: false, submitEnabled: !!form.getFieldValue('projectName') }),
```

Comment:
Why do you need to duplicate the form state into the observable? If your goal is just to call a submit with the form data could you instead pass the form into the hook or have the form register itself using context?

Reply:
I think if you pass in the form instance you can use [`form.getFieldsError()`](https://ant.design/components/form#forminstance) to check if fields have errors and disable the submit based on that.